onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,709 @@
+  }
+  return modified;
+}
+
+// SearchUpstream:
+// Recursively DFS traverse bottom-up, the graph upstream collecting all the NodeArgs that require a cast
+// inorder to move an FP16 Cast operation up the graph.
+// Visited float NodeArgs are either in require_cast or require_type_change so that the same
+// nodearg is traversed not more than once.
+static void SearchUpstream(Graph& graph, NodeArg* node_arg,
+                           std::unordered_set<NodeArg*>& require_cast,
+                           std::unordered_set<NodeArg*>& require_type_change,
+                           std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                           size_t level) {
+  Node* node = graph.GetMutableProducerNode(node_arg->Name());
+  if (node == nullptr) {
+    // The graph inputs don't have the producer nodes
+    if (IsType(*node_arg, TensorProto_DataType_FLOAT)) {
+      require_cast.insert(node_arg);
+    }
+  } else if (std::find(removed_nodes.begin(), removed_nodes.end(), node->Index()) == removed_nodes.end()) {
+    if (IsCastTo(node, TensorProto_DataType_FLOAT)) {
+      // This Cast node and the Cast node that will be created later will cancel out
+      require_cast.insert(node_arg);
+    } else {
+      std::string op_type = node->OpType();
+      if (!IsFP16Allow(op_type, level)) {
+        // Cannot traverse-up beyond this point
+        if (node_arg->Exists() && IsType(*node_arg, TensorProto_DataType_FLOAT)) {
+          require_cast.insert(node_arg);
+        }
+      } else {
+        // If the node has other float32 output(s) then stop the search.
+        for (const auto* output_def : node->OutputDefs()) {
+          // TODO: If the specified optimization is greater than 1 then insert a Cast to the
+          // other output_def and still propagate FP16 cast up the graph.
+          if (output_def != node_arg) {
+            if (IsType(*output_def, TensorProto_DataType_FLOAT)) {
+              require_cast.insert(node_arg);
+              return;
+            }
+          }
+        }
+        for (NodeArg* node_input : node->MutableInputDefs()) {
+          if (IsType(*node_input, TensorProto_DataType_FLOAT) &&
+              require_cast.find(node_input) == require_cast.end() &&
+              require_type_change.find(node_input) == require_type_change.end()) {
+            SearchUpstream(graph, node_input, require_cast, require_type_change, removed_nodes, level);
+            if (require_cast.find(node_input) == require_cast.end()) {
+              require_type_change.insert(node_input);
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+// SearchDownstream:
+// Recursively DFS traverse the graph downstream collecting all the NodeArgs that require a cast
+// inorder to remove an FP32 Cast operation up the graph. Also collect the NodeArgs that need to
+// be converted from float to float16 along the way.
+// The recursion only traverses an
+static void SearchDownstream(Graph& graph, NodeArg* node_arg,
```

Comment:
>SearchDownstream [](start = 12, length = 16)

we should log which node we have propagated over in this function. 

At the end of day, a developer will use this log to check if a node has been place correctly in fp16. 

Same comment applies to SearchUpStream.

Reply:
We actually don't perform any graph transformation in SearchUpStream/SearchDownStream we are only collecting, nodes that require cast and nodes the need type change, if we remove the original Cast node. The logs are generated by the PropagateForward/PropagateBackward.

Reply:
The end result of moving around Casts is that some node will be placed in fp16. 
Adding/Removing the Cast nodes is the "means", not the "goals".  Placing nodes in fp16 is our actually goal. 

We should log the end result for better debugging.

Reply:
Added the logs in ChangeTypeToFP16.